### DISCUSSION TITLE
[Feature Request] unique keys (short urls) for the same destination

### DISCUSSION THREAD

**Discussion by natterstefan** [dubinc/dub#408] [Ideas]
Hi @steven-tey,

as already stated via email, I want to ask you and the collaborators (thanks for all the work you invested into this project) to consider adding an option to create unique keys for the same destination.

In our case and for our business model, having n-short links for the same ((exact) destination is counterproductive. If users share the _same_ destination, it should use the existing link (key). It's not feasible to analyze n-links for the same destination (as this is also not possible right now due to the lack of a global dashboard; see #278).

That is why I request to add a `dedupe` or another option when creating a link to ensure it uses existing keys instead. One can then decide if duplicates are valid on a destination-by-destination basis.

There is a workaround you suggested (see #404 why that does not work), and this also adds an overhead when creating new links as I have to fetch first if a key exists, and only then can I create a new link or use an existing one.

Thank you.

**Comment by steven-tey** [comment]
Hey Stefan! Thanks for the feature request. To be completely honest, this sounds like a very unique (pun intended) use case to me that I'm not sure should be added to a link management tool like Dub (to avoid feature bloat).

I'll convert this into a discussion issue for now and see if other folks have the same requirements, before making a decision about adding this to Dub. 🙏

**Comment by natterstefan** [comment]
@steven-tey two more things:

## Performance
If the API already takes care of deduplication apps (like ours) do not access the API twice (first check if the destination exists) and only then create the short link. This adds a noticeable overhead (in terms of ms until one gets the result) to on-demand short link creation in the client (browser).

## Comparison with other platforms

> this sounds like a very unique (pun intended) use case to me that I'm not sure should be added to a link management tool like Dub

I just validated what one of the big players (e.g. bitly) does, and they apply what I ask for in this feature request. I don't know if this is as unique as you said.

<img width="1714" alt="image" src="https://github.com/steven-tey/dub/assets/1043668/ce0e2b0d-12ff-4981-abf3-820fe75d7a52">

Other public (=no account required) do not deduplicate:

<img width="1327" alt="image" src="https://github.com/steven-tey/dub/assets/1043668/c4bd7c05-879c-492b-bc47-5957583dc8c7">

Also tiny.url dedupes during the same session for anonymous users:

<img width="1677" alt="image" src="https://github.com/steven-tey/dub/assets/1043668/4129f551-173c-4d92-a77f-663ac13ff9c5">


IMHO the key probably is whether you need an account or not:

- public short link creation: no deduplication
- account required: with duplication turned on per default
	- but maybe with dub the account owner can decide if they want to disable it (per link creation or on for all)

And dub.co requires an account which would mean (at least in my opinion) that deduplication should be the default behavior and one should be able to turn it off instead.

I understand that customers like Hashnode (one of dub.co, correct?) _probably_ don't care. They provide it as a tool for users to easily share short-urls. But companies with dedicated strategies about how to and when to use short links might want to dedupe them to have one per unique destination.

Let's see if others agree or disagree with me :D

**Reply by steven-tey** [reply]
Really appreciate the extra context here, thank you @natterstefan!

**Reply by natterstefan** [reply]
@steven-tey you're welcome. In my opinion, context is essential to make a decision here. :)

**Comment by patrickmast** [comment]
Hello,

I'm building an internal tool that makes a short-link to a public link to an invoice of our CRM app. For this I use the new API (Which i ❤️).

Adding an option in the API to prevent creating unique keys for the same destination can help me to prevent creating a link for a long-url that already exists.

But first we need a fix for https://github.com/steven-tey/dub/issues/404 as this is a show-stopper for me. 

Thank you!

Patrick
